### PR TITLE
Refactor remove go core flag by cobra

### PR DESCRIPTION
**Pull request by aperezg**
Refactor to start using cobra, with the refactor some breaking changes are producing on the `cli`, the way to use the `cli` will be different with this version.

before the `cli` use was:
```sh
killgrave -h
Usage of killgrave:
  -config string
    	path with the configuration file
  -host string
    	if you run your server on a different host (default "localhost")
  -imposters string
    	directory where your imposters are saved (default "imposters")
  -port int
    	port to run the server (default 3000)
  -version
    	show the version of the application
```

After apply my changes the cli will be:
```sh
killgrave http -h
Configure a HTTP mock server based on your imposters

Usage:
  killgrave http [flags]

Flags:
  -h, --help           help for http
  -H, --host string    Set a different host than localhost (default "localhost")
  -P, --port int       Port to run the server (default 3000)
  -p, --proxy string   Proxy mode, the options are all, missing or none (default "none")
  -u, --url string     The url where your proxy will be call
  -w, --watcher        File watcher will reload the server with each file changes

Global Flags:
  -c, --config string      Path to your configuration file
  -i, --imposters string   Directory where your imposters are located (default "imposters")
```

So the big change was that to run the server we need to use the command `http`, the idea of that is that if in the future we will implement the `gRPC` mock could run on a different command, ie. `killgrave grpc`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
 # Changelog
 
+## v0.5.0 (TBD)
+* Remove of use of `go flags` by `cobra`
```
Comment (joanlopez) [inline CHANGELOG.md:4]:
```suggestion
* Replace the use of `go flags` by `cobra`
```

File: README.md [hunk 2]
```diff
@@ -54,38 +54,36 @@ Or you can download the binary for your arch on:
...
 
-Use `killgrave` with default flags:
+Use `killgrave http` with default flags:
 
 ```sh
 $ killgrave
 2019/04/14 23:53:26 The fake server is on tap now: http://localhost:3000
 ```
-Or custom your server with these flags:
+Or custom your server:
```
Comment (joanlopez) [inline README.md:71]:
```suggestion
Or customize your server:
```

File: README.md [hunk 3]
```diff
@@ -54,38 +54,36 @@ Or you can download the binary for your arch on:
...
+Usage:
+  killgrave http [flags]
+
+Flags:
+  -h, --help           help for http
+  -H, --host string    Set a different host than localhost (default "localhost")
+  -P, --port int       Port to run the server (default 3000)
+  -p, --proxy string   Proxy mode, the options are all, missing or none (default "none")
+  -u, --url string     The url where your proxy will be call
+  -w, --watcher        File watcher will reload the server with each file changes
```
Comment (joanlopez) [inline README.md:82]:
```suggestion
  -w, --watcher        File watcher will reload the server on each file change
```

File: README.md [hunk 4]
```diff
@@ -54,38 +54,36 @@ Or you can download the binary for your arch on:
...
+
+Global Flags:
```
Comment (joanlopez) [inline README.md:84]: I miss `-v` (version) and (global) `-h` (help) global flags.

File: internal/app/cmd/http/http.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+const (
+	_defaultHost      = "localhost"
+	_defaultPort      = 3000
+	_defaultProxyMode = killgrave.ProxyNone
+)
+
+var (
+	errGetDataFromImpostersFlag = errors.New("error trying to get data from imposters flag")
+	errGetDataFromHostFlag      = errors.New("error trying to get data from host flag")
+	errGetDataFromPortFlag      = errors.New("error trying to get data from Port flag")
```
Comment (joanlopez) [inline internal/app/cmd/http/http.go:29]:
```suggestion
	errGetDataFromPortFlag      = errors.New("error trying to get data from port flag")
```

File: internal/app/cmd/http/http.go [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return runHTTP(cmd, cfg)
+		},
+		Args: cobra.NoArgs,
+	}
+
+	cmd.PersistentFlags().StringP("host", "H", _defaultHost, "Set a different host than localhost")
+	cmd.PersistentFlags().IntP("port", "P", _defaultPort, "Port to run the server")
+	cmd.PersistentFlags().BoolP("watcher", "w", false, "File watcher will reload the server with each file changes")
```
Comment (joanlopez) [inline internal/app/cmd/http/http.go:57]:
```suggestion
	cmd.PersistentFlags().BoolP("watcher", "w", false, "File watcher will reload the server on each file change")
```

File: README.md [hunk 7]
```diff
@@ -54,38 +54,36 @@ Or you can download the binary for your arch on:
...
-        file watcher, reload the server with each file change
+Usage:
+  killgrave http [flags]
+
+Flags:
+  -h, --help           help for http
+  -H, --host string    Set a different host than localhost (default "localhost")
+  -P, --port int       Port to run the server (default 3000)
+  -p, --proxy string   Proxy mode, the options are all, missing or none (default "none")
+  -u, --url string     The url where your proxy will be call
```
Comment (joanlopez) [inline README.md:81]:
```suggestion
  -u, --url string     The url where the proxy will redirect to
```

File: internal/app/cmd/http/http.go [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+	cmd.Flags().StringP("proxy", "p", _defaultProxyMode.String(), "Proxy mode, the options are all, missing or none")
+	cmd.Flags().StringP("url", "u", "", "The url where your proxy will be call")
```
Comment (joanlopez) [inline internal/app/cmd/http/http.go:59]:
```suggestion
	cmd.Flags().StringP("url", "u", "", "The url where the proxy will redirect to")
```

File: internal/config.go [hunk 9]
```diff
@@ -91,57 +99,56 @@ func (p *ProxyMode) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+func NewConfig(impostersPath, host string, port int) (Config, error) {
+	if impostersPath == "" {
+		return Config{}, errEmptyImpostersPath
+	}
+
+	if host == "" {
+		return Config{}, errEmptyHost
+	}
+
+	if port <= 0 {
```
Comment (joanlopez) [inline internal/config.go:121]:
```suggestion
	if port < 0 || port > 65535 {
```

Comment (joanlopez) [reply internal/config.go:121]: According to [RFC 793](https://tools.ietf.org/html/rfc793), the port is a 16-bit unsigned int.

This means the range is 0 - 65535.

However, in the near future, we could potentially add some additional checks (reserved ports, ephemeral ports, etc. However, the range check is the least restrictive check, I guess.

File: internal/config.go [hunk 10]
```diff
@@ -91,57 +99,56 @@ func (p *ProxyMode) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+	if err != nil {
+		return Config{}, fmt.Errorf("%w: error trying to read config file: %s, using default configuration instead", err, cfgPath)
+	}
+	defer configFile.Close()
 
-		return nil
+	var cfg Config
+	bytes, _ := ioutil.ReadAll(configFile)
+	if err := yaml.Unmarshal(bytes, &cfg); err != nil {
+		return Config{}, fmt.Errorf("%w: error while unmarshall configFile file %s, using default configuration instead", err, cfgPath)
```
Comment (joanlopez) [inline internal/config.go:148]:
```suggestion
		return Config{}, fmt.Errorf("%w: error while unmarshalling configFile file %s, using default configuration instead", err, cfgPath)
```

**Review by joanlopez — APPROVED** [review]
LGTM!! 🚀 

PS: I left few minor comments, feel free to fix them and merge it.

File: README.md (hunk 4 above)
Comment (aperezg) [reply README.md:84]: Sorry the `-v` option is only on the root command `killgrave` this is the `http` command that for now is the only that we have.

And the help is there on the flags :P

File: internal/config.go (hunk 9 above)
Comment (aperezg) [reply internal/config.go:121]: or using directly a uint ^^'

**Pull request merged**